### PR TITLE
executor: refine split index region

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs520 <crazycs520@gmail.com>

### What problem does this PR solve?

Fix https://github.com/pingcap/tidb/issues/17041.

When `split region` for the first index, no need to split in the index start key, it's useless, because the table has been splitted when the table been created. 

### What is changed and how it works?

here is an example:

suppose the table is:  `create table t (a int, index idx(a))`;

when create table, the table region is as below:

```sql
region1: [tid_, nextTid_)  -- tid is the table id, nextTid is the next table id.

```

after execute `split table t index idx between (0) and (20000) regions 2;` 

before this PR, the table region will be:

```sql
region1:  [tid_,          tid_i_1_)  -- actually the region is useless, no key will write into this region.
region2:  [tid_i_1_,      tid_i_1_10000)   
region3:  [tid_i_1_10000, nextTid_)  -- the index data that `a > 10000` and the record data will write into this region, this is not expect, the record data should write into a separate region. 
```

And since the `region1` is useless, when you execute `show table t regions`, you won't find the `region1` in the result.

This PR, after execute split index, the table region will be:

```sql
region1:  [tid_,      tid_i_1_10000)   -- 1 is the index `idx` id.
region2:  [tid_i_1_10000, tid_i_2_)    -- 2is the next index id.
region3:  [tid_i_2,       nextTid_])

```

The `region1` and `region2` will used to store index `idx` data, and the `region3` now used to store record data.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 
- Split index that no need to split the start index key when the index is the first index.

### CODE REVIEW

**Comment by crazycs520** [comment]
@XuHuaiyu @zimulala PTAL

File: executor/split.go [hunk 1]
```diff
@@ -172,19 +172,16 @@ func (e *SplitIndexRegionExec) getSplitIdxPhysicalKeysFromValueList(physicalID i
...
-		break
+	// 1. Split in the start key for the index if the index is not the first index.
+	// For the first index, split the start key is useless.
+	if len(e.tableInfo.Indices) > 0 && e.tableInfo.Indices[0].ID != e.indexInfo.ID {
+		startKey := tablecodec.EncodeTableIndexPrefix(physicalID, e.indexInfo.ID)
+		keys = append(keys, startKey)
 	}
+
+	// 2. Split in the end key.
+	endKey := tablecodec.EncodeTableIndexPrefix(physicalID, e.indexInfo.ID+1)
```
Comment (zimulala) [inline executor/split.go:183]: After this PR, it seems we must have `endKey`, which is different from the old code. Why change it, and do we have a test for this code.

Comment (crazycs520) [reply executor/split.go:183]: suppose the table is:  `create table t (a int, index idx(a))`;
then execute `split table t index idx between (0) and (20000) regions 2;` 

before this PR, the table region will be:

```sql
region1:  [tid_,          tid_i_1_)  
region2:  [tid_i_1_,      tid_i_1_10000)   
region3:  [tid_i_1_10000, nextTid_)  
```

As you can see,  the index data that `a > 10000` and the record data will write into region3, this is not expected, the record data should write into a separate region.

Comment (zimulala) [reply executor/split.go:183]: https://github.com/pingcap/tidb/pull/17070/files#diff-6d9a596d0d155b4342a598ba359efc8bR4494 is the test.

Comment (crazycs520) [reply executor/split.go:183]: Already has unit test for this.

Comment (zimulala) [reply executor/split.go:183]: Got it.

**Pull request merged**